feat(frontend): add custom reading settings and email subscription pages and logout integration

ericsen-tsai opened the pull request:
## Summary

This PR introduces two new membership pages: a custom reading settings page and an email subscription page. It also adds authentication middleware for route protection, logout functionality, and reusable UI components (Checkbox and Switch). The implementation includes optimistic updates for member reading settings with debounced API calls to improve user experience.

## Changes

- **New Pages:**
  - Added `/custom` page for reading settings (Baodaozai toggle and essay question count configuration)
  - Added `/email-subscription` page for newsletter subscription with preview and subscription links

- **New Components:**
  - Created `Checkbox` component (radio-style checkbox for essay question count selection)
  - Created `Switch` component (toggle switch for Baodaozai enable/disable)

- **Authentication & Security:**
  - Implemented Next.js middleware for protected route authentication
  - Added `/api/logout` endpoint for secure logout with cookie clearing
  - Enhanced auth store with logout functionality

- **Member Profile Management:**
  - Created `use-optimistic-update-member-profile` hook with debounced API calls
  - Added `use-debounce-value` hook for value debouncing
  - Updated member profile mutations to support `showBaodaozai` and `essayQuestionCount` fields
  - Enhanced side menu with logout button functionality

- **Assets & Constants:**
  - Added newsletter icon SVG (`newsletter-icon.svg`)
  - Added new route constants for custom and email-subscription pages
  - Added newsletter preview and subscription URL constants

- **UI/UX Improvements:**
  - Updated side menu to use button-based logout instead of link
  - Added responsive design support for new pages
  - Implemented optimistic UI updates for immediate feedback

## Additional Notes

- The optimistic update hook uses debouncing (1000ms) to batch API calls and reduce server load
- Protected routes are automatically validated via middleware before rendering
- Logout functionality clears authentication cookies and redirects to home page
- All new pages follow the existing membership page layout pattern with side menu integration
- This PR is based on #837 

## Screenshot(s)

<img width="1478" height="693" alt="image" src="https://github.com/user-attachments/assets/01096f89-ac67-454f-9e18-9a90a950b5bf" />

https://github.com/user-attachments/assets/ab5cc062-89e6-4a5e-a33f-fa4d25f62601
## Reference(s)

- [Notion Page 1](https://www.notion.so/twreporter/UI-2a28dbdca93280f48fd5f61182fd188d?v=1588dbdca93281dfa43c000c8415265e&source=copy_link)
- [Notion Page 2](https://www.notion.so/twreporter/26b8dbdca9328029b6eddab6b347168c?v=1588dbdca93281dfa43c000c8415265e&source=copy_link)
- [Figma Design](https://www.figma.com/design/FqWHr59xE8l9teNoCmJGVn/%E6%9C%83%E5%93%A1%E9%A0%81?node-id=462-119676&m=dev)

ericsen-tsai committed:
feat(frontend): add checkbox and switch components

ericsen-tsai committed:
feat(frontend): add custom page

ericsen-tsai committed:
feat(frontend): add email subscription module and newsletter icon

ericsen-tsai committed:
feat(frontend): implement authentication middleware and logout functionality

- Added middleware to protect routes and validate ID tokens.
- Created logout API endpoint to clear authentication cookies.
- Updated side menu to handle logout action with error handling.
- Introduced constants for protected routes and logout endpoint.

ericsen-tsai committed:
feat(frontend): enhance member profile management with new fields and optimistic updates

- Added `showBaodaozai` and `essayQuestionCount` fields to the member profile GraphQL schema.
- Updated the member profile mutation to handle new fields and improved the mutation hook with options.
- Introduced a custom hook for optimistic updates to member reading settings, allowing for immediate UI feedback.
- Implemented debouncing for reading settings updates to optimize performance.
- Refactored components to utilize new fields and hooks for better state management.

ericsen-tsai committed:
feat(frontend): enhance middleware and component functionality

- Added timeout handling in the authentication middleware to improve request reliability.
- Updated Checkbox component to change the onChange handler to accept a string value instead of a boolean.
- Added aria-label to the Switch component for better accessibility.
- Removed unnecessary class names from disabled input fields in the EditMode component for cleaner UI.

ericsen-tsai commented on packages/frontend/src/components/checkbox.tsx:
```diff
@@ -0,0 +1,49 @@
+        value={value}
+        name="questionCount"
+        checked={checked}
+        onChange={() => onChange(!checked)}
```
Fixed in [feat(frontend): enhance middleware and component functionality](https://github.com/kids-reporter/kids-reporter-monorepo/pull/838/commits/2232295c859cdb678247c66a154f631da78b28ce)

ericsen-tsai commented on packages/frontend/src/modules/membership/account/index.tsx:
```diff
@@ -0,0 +1,242 @@
+  const handleUpdateMemberProfile = useCallback(
+    async (data: AccountFormData) => {
+      try {
+        const avatarUrlFieldState = getFieldState('avatarUrl')
+        const isAvatarDirty = avatarUrlFieldState?.isDirty
+        let avatarId: string | undefined
+        if (isAvatarDirty && avatarFileRef.current) {
+          const oldAvatarId = member?.avatar?.id
+          if (oldAvatarId) {
+            await deletePhoto(oldAvatarId)
+          }
+          const newAvatar = await uploadPhoto({
+            file: avatarFileRef.current,
+            fileName: data.name ?? '',
+          })
+          avatarId = newAvatar?.id
+          avatarFileRef.current = null
+        }
```
Fixed in [feat(frontend): enhance middleware and component functionality](https://github.com/kids-reporter/kids-reporter-monorepo/pull/838/commits/2232295c859cdb678247c66a154f631da78b28ce)

ericsen-tsai commented on packages/cms/lists/photo.ts:
```diff
@@ -90,8 +92,12 @@ const listConfigurations: ListConfig<any> = list({
           Object.entries(resizedTargets).forEach(([key, value]) => {
+            const resizedFilename =
+              envVar.nodeEnv !== 'production'
+                ? `${filename}${extension}`
```
Adjusted in the last PR

ericsen-tsai commented on packages/frontend/src/modules/membership/account/index.tsx:
```diff
@@ -0,0 +1,242 @@
+
+        await updateMemberProfile({
+          data: {
+            ...(isAvatarDirty
+              ? { avatar: { connect: { id: avatarId ?? '' } } }
+              : {}),
```
In current scenario, user always uploads avatar, so when the field is dirty, there's always a value for it

ericsen-tsai commented on packages/frontend/src/components/switch.tsx:
```diff
@@ -0,0 +1,28 @@
+}
+
+function Switch({ checked, onChange }: SwitchProps) {
+  return (
+    <button
+      onClick={() => onChange(!checked)}
+      className={cn(
+        'relative flex h-5 w-10 cursor-pointer items-center rounded-full p-0.5 transition-colors duration-100 ease-in-out',
+        checked ? 'bg-red-400' : 'bg-neutral-300'
+      )}
+      role="switch"
+      aria-checked={checked}
+    >
```
Fixed in [feat(frontend): enhance middleware and component functionality](https://github.com/kids-reporter/kids-reporter-monorepo/pull/838/commits/2232295c859cdb678247c66a154f631da78b28ce)

ericsen-tsai commented on packages/frontend/src/modules/membership/account/edit-user-avatar.tsx:
```diff
@@ -0,0 +1,116 @@
+    if (!file) {
+      return
+    }
+    if (!ALLOWED_IMAGE_TYPES.includes(file.type)) {
+      setError('avatarUrl', {
+        message: '不支援的檔案格式',
+      })
+      return
+    }
+    if (file.size > MAX_IMAGE_SIZE) {
+      setError('avatarUrl', {
+        message: '檔案大小超過限制，請上傳小於 5MB 的圖片',
+      })
+      return
```
User needs to upload another file to update the file successfully, so the handle should be okay right now

ericsen-tsai commented on packages/frontend/src/modules/membership/custom/hooks/use-optimistic-update-member-profile.ts:
```diff
@@ -0,0 +1,169 @@
+    handleUpdateMemberReadingSettings({
+      showBaodaozai: debouncedShowBaodaozai,
+      essayQuestionCount: debouncedEssayQuestionCount,
+    })
```
I think this is OK to update all field by once

ericsen-tsai commented on packages/routing-ui/src/components/input.tsx:
```diff
@@ -132,13 +150,18 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
 
     return (
       <div className="gap-2 flex flex-col">
         <div className={inputClasses} ref={ref}>
```
The relative classname is in the inputClass

ericsen-tsai commented on packages/api-gateway/src/gql-proxy-mini-app.ts:
```diff
@@ -208,12 +208,19 @@ export function createGraphQLProxy({
+        // Preserve original Content-Type for multipart/form-data (file uploads)
+        // Only set to application/json for regular GraphQL requests
+        const originalContentType = req.get('Content-Type') || ''
+        if (originalContentType.includes('multipart/form-data')) {
+          // Preserve multipart/form-data with boundary for file uploads
+          proxyReq.setHeader('Content-Type', originalContentType)
+        } else {
+          // Set to application/json for regular GraphQL JSON requests
+          proxyReq.setHeader('Content-Type', 'application/json')
+        }
```
Since current scenario only handle image file upload, this should be OK

ericsen-tsai commented on packages/frontend/src/modules/membership/types.ts:
```diff
@@ -1,8 +1,18 @@
+export const accountFormSchema = z.object({
+  name: z.string().min(1, '請輸入全名'),
+  nickname: z.string().optional(),
+  contactEmail: z.union([z.email('請輸入有效的電子郵件格式'), z.literal('')]),
```
Email is valid in zod 4
https://zod.dev/api?id=emails

ericsen-tsai commented on packages/frontend/src/modules/membership/account/index.tsx:
```diff
@@ -0,0 +1,242 @@
+          const oldAvatarId = member?.avatar?.id
+          if (oldAvatarId) {
+            await deletePhoto(oldAvatarId)
+          }
```
This is handled in the previous PR

ericsen-tsai commented on packages/frontend/src/app/api/logout/route.ts:
```diff
@@ -0,0 +1,24 @@
+
+  response.cookies.set(ID_TOKEN_COOKIE_NAME, '', {
+    httpOnly: true,
+    secure: process.env.NODE_ENV === 'production',
```
I think current method is OK for testing, if there's further improvement, I'll add one more env

ericsen-tsai commented on packages/frontend/src/api/member.ts:
```diff
@@ -58,3 +62,28 @@ export const getMemberProfileByMemberId = async ({
+    }
+  )
+
+  return response?.data?.data
```
Fixed in the previous PR

ericsen-tsai commented on packages/frontend/src/modules/membership/account/edit-mode.tsx:
```diff
@@ -0,0 +1,139 @@
+              value={id}
+              placeholder={id}
+              disabled
+              className="w-full rounded-xl border border-neutral-400 bg-neutral-100 desktop:bg-neutral-100!"
```
Fixed in [fix(frontend): enhance input component style](https://github.com/kids-reporter/kids-reporter-monorepo/pull/838/commits/d4927c52ad769ae69b0733d0b5a6b6df26546705)

ericsen-tsai commented on packages/frontend/src/middleware.ts:
```diff
@@ -0,0 +1,63 @@
+    )
+  } catch {
+    return false
+  }
```
Fixed in [feat(frontend): enhance middleware and component functionality](https://github.com/kids-reporter/kids-reporter-monorepo/pull/838/commits/2232295c859cdb678247c66a154f631da78b28ce)

ericsen-tsai committed:
fix(frontend): move redirection after logout success

ericsen-tsai commented on packages/frontend/src/modules/membership/components/side-menu.tsx:
```diff
@@ -14,6 +17,19 @@ function MembershipSideMenu() {
+  const { logout } = useHydratedAuthStore()
+  const router = useRouter()
+  const handleLogout = async () => {
+    try {
+      await logout()
+    } catch {
+      toast.error('登出失敗')
+    } finally {
+      router.push('/')
+    }
```
Fixed in [fix(frontend): move redirection after logout success](https://github.com/kids-reporter/kids-reporter-monorepo/pull/838/commits/a1bb8412aa673db89ad3220bbf80a2812bd7e3cb)

nickhsine commented on packages/frontend/src/app/(membership)/custom/page.tsx:
```diff
@@ -0,0 +1,7 @@
+
+function CustomPage() {
+  return <CustomModule />
+}
```
nitpick:
是不是直接 `export default CustomModule` 就好？

ericsen-tsai replied:
Yes, but I think I'll leave it as it is for two reasons
1. consistency in other pages
2. if some server call needs to be implemented some day, it will be added in this layer

nickhsine commented on packages/frontend/src/app/(membership)/email-subscription/page.tsx:
```diff
@@ -0,0 +1,7 @@
+
+function EmailSubscriptionPage() {
+  return <EmailSubscriptionModule />
+}
```
nitpick:
直接 `export default EmailSubscriptionModule`?

ericsen-tsai replied:
Replied in https://github.com/kids-reporter/kids-reporter-monorepo/pull/838#discussion_r2570355598

nickhsine commented on packages/frontend/src/modules/membership/components/side-menu.tsx:
thought:
會員相關的服務和頁面都採取 client side rendering 的話，我們會需要保護 `/custom`, `/myreading` 等 endpoint 嗎？
如果使用者沒有登入，是不是要呈現尚未登入的提示，讓他登入後再使用？

ericsen-tsai replied:
Yeah, currently user only redirect to index page without any hints, which is bad UX.
I've modified the middleware logic to redirect to login page with destination and toast hint.
[fix(frontend): enhance login functionality and middleware redirection](https://github.com/kids-reporter/kids-reporter-monorepo/pull/838/commits/37841e8e4963bc8f5aeb26cc45e35341c455e888)

ericsen-tsai replied:
Changed to check on client side
[refactor(frontend): remove legacy middleware and implement AuthRouteGuard for route protection](https://github.com/kids-reporter/kids-reporter-monorepo/pull/838/commits/58fa05599f76c6b6c7ec36ddf877ccedac4ae162)

nickhsine commented on packages/frontend/src/modules/membership/custom/index.tsx:
```diff
@@ -1,16 +1,18 @@
-
+  const { member } = useHydratedAuthStore()
+  const { optimisticUpdateMemberReadingSettings } =
+    useOptimisticUpdateMemberReadingSettings()
```
thought:

我們可以讓 component 用 state，例如 `showBaodaozai` state， 來存使用者在 UI 上看到的狀態，當使用者快速 toggle `showBaodaozai` 時，我們透過 setState 來 re-render UI，再透過 debounce 的方式，等使用者 toggle 後一陣子，再發送 GQL request ，來達成 optimistic update 的效果。

auth store 裡的資料則採用 single source of truth ，統一和資料庫的資料 sync，UI 的 state 不會影響到 auth store 裡的資料，減低 race condition 的問題。

ericsen-tsai replied:
Done [refactor(frontend): enhance member profile management with abort signal handling](https://github.com/kids-reporter/kids-reporter-monorepo/pull/838/commits/f0daba1b8ad3691364c55919f32da462910f3d61), muchas gracias

ericsen-tsai committed:
fix(frontend): enhance login functionality and middleware redirection

- Updated middleware to include a destination parameter in the redirection URL after failed authentication.
- Refactored login page to utilize a new Login component, improving code organization and readability.
- Enhanced destination sanitization to return an object with destination and subpath information.
- Integrated toast notifications for login prompts based on the destination path.

ericsen-tsai committed:
refactor(frontend): enhance member profile management with abort signal handling

- Added optional `abortSignal` parameter to the `getMemberProfileByMemberId` function for better request management.
- Updated the `useAuthStore` to handle aborting previous fetch requests, preventing race conditions.
- Refactored the `useOptimisticUpdateMemberReadingSettings` hook to manage local state for `showBaodaozai` and `essayQuestionCount`.
- Enhanced the `Custom` component to utilize local state for reading settings.

ericsen-tsai committed:
refactor(frontend): remove legacy middleware and implement AuthRouteGuard for route protection

- Deleted the existing middleware file to streamline authentication handling.
- Introduced AuthRouteGuard component to manage authentication state and redirect unauthenticated users.
- Updated layout to wrap children with AuthRouteGuard, ensuring protected routes are enforced.

nickhsine commented:
LGTM 🥇 🥇 
Nice work!

nickhsine commented on packages/frontend/src/constants/route.ts:
nitpick:
remove `PROTECTED_ROUTES` since it is not used.

ericsen-tsai committed:
fix(frontend): remove route.ts
